### ISSUE TITLE
False positive on SC2016 -- or maybe an interesting corner case?

### ISSUE THREAD

**Issue by tspearconquest** [koalaman/shellcheck#2556]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2016
- My shellcheck version (`shellcheck --version` or "online"): 0.8.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh

set -e

ME=$(basename $0)

auto_envsubst() {
  local template_dir="${NGINX_ENVSUBST_TEMPLATE_DIR:-/etc/nginx/templates}"
  local suffix="${NGINX_ENVSUBST_TEMPLATE_SUFFIX:-.template}"
  local output_dir="${NGINX_ENVSUBST_OUTPUT_DIR:-/etc/nginx/conf.d}"

  local template defined_envs relative_path output_path subdir
  defined_envs=$(printf '${%s} ' $(env | cut -d= -f1))
  [ -d "$template_dir" ] || return 0
  if [ ! -w "$output_dir" ]; then
    echo >&3 "$ME: ERROR: $template_dir exists, but $output_dir is not writable"
    return 0
  fi
  find "$template_dir" -follow -type f -name "*$suffix" -print | while read -r template; do
    relative_path="${template#$template_dir/}"
    output_path="$output_dir/${relative_path%$suffix}"
    subdir=$(dirname "$relative_path")
    # create a subdirectory where the template file exists
    mkdir -p "$output_dir/$subdir"
    echo >&3 "$ME: Running envsubst on $template to $output_path"
    envsubst "$defined_envs" < "$template" > "$output_path"
  done
}

auto_envsubst

exit 0
```

#### Here's what shellcheck currently says:
```
In 1.22.0/20-envsubst-on-templates.sh line 13:
  defined_envs=$(printf '${%s} ' "$(env | cut -d= -f1)")
                        ^------^ SC2016 (info): Expressions don't expand in single quotes, use double quotes for that.
```

#### Here's what I wanted or expected to see:
Hello!

This expression is intentionally single quoted to prevent expansion until after the nested subshell completes and printf expands %s. When printf outputs what it needs to, it's outputting a proper variable which is evaluated by the first subshell in order to set `defined_envs`.

This appears to be a really clever way to avoid using `eval` to populate the variable. I found this in one of the build scripts for the `nginx` docker image: https://github.com/nginxinc/docker-nginx/blob/master/stable/alpine/20-envsubst-on-templates.sh#L13

I feel that this is an appropriate enough use case to want to propose that this be added to shellcheck so that it doesn't trigger SC2016, so I realize I'm ignoring the warning below by submitting this issue, and hope you'll consider the above.

> However, there's a long tail of less common commands and flags that also frequently expect $s, and it's not in ShellCheck's scope to try to keep track of them all. When you come across such a command, please [ignore](https://www.shellcheck.net/wiki/ignore) the suggestion, either permanently or for that one instance.

Thoughts?

**Comment by hyperupcall** [comment]
Definitely a bug. In my case, I get:

```text
In themes/iterate/iterate.theme.bash line 49:
        new_prompt=$(PS1="$new_PS1" "$BASH" --norc -i < /dev/null 2>&1 | sed -n '${s/^\(.*\)exit$/\1/p;}')
                                                                                ^-----------------------^ SC2016 (info): Expressions don't expand in single quotes, use double quotes for that.
```

If the contents isn't a bash expression, no error should show.

**Comment by djmattyg007** [comment]
I've encountered this with just the most basic use of `envsubst`:

```bash
envsubst '$MYVAR' < file-that-contains-variables.txt | other-program-that-consumes-input
```

**Comment by wileyhy** [comment]
Possibly the '${'...'}' syntax of the sed script is triggering the false
positive, but an example of similar syntax is included on "page 36" of
sed's GNU manual:

  $ sed -n 's/.//g ; H ; ${x;s/\n//g;l}' invalid.txt
  \316\316$

OP, I don't know what your goal for this code might be, but (I wonder...)
you could be looking for ANSI C quoting...?

~ $ declare -p PS1 declare -- PS1="\\[\\e[0;32m\\]\\w\\[\\e[0m\\]
\\[\\e[0;97m\\]\\\$\\[\\e[0m\\] "
~ $ old_PS1=$PS1
~ $ new_PS1=$' \\[\\e[0;97m\\]\\\$\\[\\e[0m\\] '
~ $ PS1=$new_PS1
 $ PS1=$old_PS1
~ $ declare -p PS1 {new,old}_PS1 declare --
PS1="\\[\\e[0;32m\\]\\w\\[\\e[0m\\] \\[\\e[0;97m\\]\\\$\\[\\e[0m\\] "
declare -- new_PS1=" \\[\\e[0;97m\\]\\\\\$\\[\\e[0m\\] "
declare -- old_PS1="\\[\\e[0;32m\\]\\w\\[\\e[0m\\]
\\[\\e[0;97m\\]\\\$\\[\\e[0m\\] "

Wiley

**Comment by hyperupcall** [comment]
SC2016 has several issues, I've run into #2693 as well
